src/levenshtein.cr

```diff
@@ -13,36 +13,68 @@ module Levenshtein
   def self.distance(string1 : String, string2 : String) : Int32
     return 0 if string1 == string2
 
-    s = string1.chars
-    t = string2.chars
-
-    s_size = s.size
-    t_size = t.size
+    s_size = string1.size
+    t_size = string2.size
 
     return t_size if s_size == 0
     return s_size if t_size == 0
 
-    # This is to allocate less memory
-    if t_size > s_size
-      t, s = s, t
-      t_size, s_size = s_size, t_size
-    end
+    if string1.ascii_only? && string2.ascii_only?
+      s = string1.to_unsafe
+      t = string2.to_unsafe
 
-    v = Pointer(Int32).malloc(t_size + 1) { |i| i }
+      # This is to allocate less memory
+      if t_size > s_size
+        t, s = s, t
+        t_size, s_size = s_size, t_size
+      end
 
-    s_size.times do |i|
-      last_cost = i + 1
+      costs = Pointer(Int32).malloc(t_size + 1) { |i| i }
 
-      t_size.times do |j|
-        sub_cost = s[i] == t[j] ? 0 : 1
-        cost = Math.min(Math.min(last_cost + 1, v[j + 1] + 1), v[j] + sub_cost)
-        v[j] = last_cost
-        last_cost = cost
+      last_cost = 0
+      s_size.times do |i|
+        last_cost = i + 1
+
+        t_size.times do |j|
+          sub_cost = s[i] == t[j] ? 0 : 1
+          cost = Math.min(Math.min(last_cost + 1, costs[j + 1] + 1), costs[j] + sub_cost)
+          costs[j] = last_cost
+          last_cost = cost
+        end
+        costs[t_size] = last_cost
       end
-      v[t_size] = last_cost
-    end
 
-    v[t_size]
+      last_cost
+    else
+      reader = Char::Reader.new(string1)
+
+      # Use an array instead of a reader to decode the second string only once
+      chars = string2.chars
+
+      # This is to allocate less memory
+      if t_size > s_size
+        chars, reader = reader, chars
+        t_size = s_size
+      end
+
+      costs = Pointer(Int32).malloc(t_size + 1) { |i| i }
+
+      last_cost = 1
```

Comment:
Why is `last_cost = 1` in this branch, when it is `0` in the other?

Reply:
Good question. I'm not quite sure anymore so I will probably change this back.